egg/zoo/dsprites_bvae/train.py

```diff
@@ -0,0 +1,210 @@
+import egg.core as core
+import pathlib
+from torch.autograd import Variable
+import json
+from typing import Union, Callable
+import os
+
+
+
+import numpy as np
+from .archs import VisualReceiver, VisualSender
+from egg.zoo.dsprites_bvae.data_loaders.data_loaders import get_dsprites_dataloader
+from egg.core.language_analysis import TopographicSimilarity, PosDisent
+
+
+class TopographicSimilarityLatents(TopographicSimilarity):
+    def __init__(self,
+                 sender_input_distance_fn: Union[str, Callable] = 'cosine',
+                 message_distance_fn: Union[str, Callable] = 'edit',
+                 compute_topsim_train_set: bool = False,
+                 compute_topsim_test_set: bool = True):
+
+        super().__init__(sender_input_distance_fn, message_distance_fn, compute_topsim_train_set, compute_topsim_test_set)
+
+    def compute_similarity(self, sender_input: torch.Tensor, messages: torch.Tensor, mode: str,  epoch: int):
+        def compute_distance(_list, distance):
+            return [distance(el1, el2)
+                    for i, el1 in enumerate(_list[:-1])
+                    for j, el2 in enumerate(_list[i+1:])
+                    ]
+
+        messages = [msg.tolist() for msg in messages]
+        input_dist = compute_distance(
+            sender_input.numpy(), self.sender_input_distance_fn)
+        message_dist = compute_distance(messages, self.message_distance_fn)
+        topsim = spearmanr(input_dist, message_dist,
+                           nan_policy='raise').correlation
+
+        output_message = json.dumps(dict(topsim=topsim, epoch=epoch))
+        print(output_message, flush=True)
+
+
+def reconstruction_loss(x, x_recon, distribution='bernoulli'):
+    batch_size = x.size(0)
+    assert batch_size != 0
+
+    if distribution == 'bernoulli':
+        recon_loss = F.binary_cross_entropy_with_logits(x_recon, x, size_average=False).div(batch_size)
+    elif distribution == 'gaussian':
+        x_recon = F.sigmoid(x_recon)
+        recon_loss = F.mse_loss(x_recon, x, size_average=False).div(batch_size)
+    else:
+        recon_loss = None
+
+    return recon_loss
+
+
+def kl_divergence(mu, logvar):
+    batch_size = mu.size(0)
+    assert batch_size != 0
+    if mu.data.ndimension() == 4:
+        mu = mu.view(mu.size(0), mu.size(1))
+    if logvar.data.ndimension() == 4:
+        logvar = logvar.view(logvar.size(0), logvar.size(1))
+
+    klds = -0.5*(1 + logvar - mu.pow(2) - logvar.exp())
+    total_kld = klds.sum(1).mean(0, True)
+    dimension_wise_kld = klds.mean(0)
+    mean_kld = klds.mean(1).mean(0, True)
+
+    return total_kld, dimension_wise_kld, mean_kld
+
+
+def reparametrize(mu, logvar):
+    std = logvar.div(2).exp()
+    eps = Variable(std.data.new(std.size()).normal_())
+    return mu + std * eps
+
+class betaVAE_Game(nn.Module):
+    def __init__(self, sender, receiver, z_dim=10, beta=4):
+        """Model proposed in the original beta-VAE paper(Higgins et al, ICLR, 2017)."""
```

Comment:
Maybe worth having it in egg/core so that it's available across egg? @eugene-kharitonov

Reply:
@robertodessi What do you call `it` here? :P

Reply:
@robertodessi Do you mean moving the convolutional agents to one of the egg/core files?

Reply:
I meant moving the betaVAE_Game class to egg/core if some other game want to use it e.g. with different archs or different data